rs/tests/networking/canister_http_correctness_test.rs

```diff
@@ -610,30 +814,1006 @@ pub fn test_http_calls_to_ic_fails(env: TestEnv) {
+                    .to_string(),
+            )
+        })
+        .collect();
+
+    // Rule 1: Check that all left headers exist in right
+    let http_bin_server_received_all_outcall_headers = request_headers
+        .iter()
+        .all(|x| http_bin_server_received_headers.contains(x));
+
+    assert!(
+        http_bin_server_received_all_outcall_headers,
+        "1. HTTP bin server did not receive all headers specified in the outcall. Specified headers: {:?}, received headers: {:?}",
+        request_headers,
+        http_bin_server_received_headers
+    );
+
+    // Rule 2: Check that all headers received by the server was specified in outcall.
+    let http_bin_server_only_received_headers_specified_by_outcall =
+        http_bin_server_received_headers
+            .iter()
+            .filter(|(name, value)| {
+                !matches!(
+                    (name.as_str(), value.as_str()),
+                    ("host", _)
+                        | ("content-length", _)
+                        | ("accept-encoding", _)
+                        | ("user-agent", "ic/1.0")
+                )
+            })
+            .all(|(name, value)| request_headers.contains(&(name.clone(), value.clone())));
+
+    assert!(http_bin_server_only_received_headers_specified_by_outcall,
+        "2. Http bin server received headers that were not specified in the outcall. Specified headers: {:?}, received headers: {:?}",
+        request_headers,
+        http_bin_server_received_headers);
+
+    let request_method = match request.method {
+        HttpMethod::GET => "GET",
+        HttpMethod::POST => "POST",
+        HttpMethod::HEAD => "HEAD",
+    };
+
+    assert_eq!(
+        request_method,
+        response_body["method"].as_str().unwrap(),
+        "3. Mismatch in HTTP method."
+    );
+
+    let server_received_body = response_body["data"].as_str().unwrap();
+    let outcall_sent_body = String::from_utf8(request.body.clone().unwrap_or_default()).unwrap();
+
+    assert_eq!(
+        server_received_body, &outcall_sent_body,
+        "4. HTTP bin server received body does not match the outcall sent body."
+    );
+}
+type ProxyCanisterResponse = Result<RemoteHttpResponse, (RejectionCode, String)>;
+type OutcallsResponse = Result<RemoteHttpResponse, RejectResponse>;
+
+async fn submit_outcall<Request>(handlers: &Handlers<'_>, request: Request) -> OutcallsResponse
+where
+    Request: Clone + CandidType,
+{
+    let args = Encode!(&request).unwrap();
+    let agent = handlers.agent().await;
+
+    let principal_id: PrincipalId = handlers.proxy_canister().effective_canister_id();
+    let principal: Principal = principal_id.into();
+
+    agent
+        .update(&principal, "send_request")
+        .with_arg(args)
+        .call_and_wait()
         .await
-        .expect("Request completes.")
+        .map_err(|agent_error| match agent_error {
+            AgentError::CertifiedReject(response) | AgentError::UncertifiedReject(response) => {
```

Comment:
```suggestion
            AgentError::CertifiedReject(response) => {
```
we should never see uncertified rejects